src/core_deku/ticket_table.ml

```diff
@@ -0,0 +1,126 @@
+open Helpers
+
+module Tickets = struct
+  module Ticket_map = Map.Make_with_yojson (Ticket_id)
+
+  type t = Amount.t Ticket_map.t [@@deriving yojson]
+
+  let get_balance t ~ticket = Ticket_map.find_opt ticket t
+
+  let of_seq = Ticket_map.of_seq
+
+  let singleton ~ticket ~amount =
+    Ticket_map.of_seq (Seq.return (ticket, amount))
+
+  let merge t ~ticket ~amount =
+    let merger option ~amount =
+      let value = Option.fold ~none:amount ~some:Amount.(( + ) amount) option in
+      value in
+    Ticket_map.update ticket
+      (fun x ->
+        let merged = merger ~amount x in
+        Option.some merged)
+      t
+
+  let get_and_remove t ~ticket ~to_take =
+    let%ok amount =
+      Ticket_map.find_opt ticket t |> Option.to_result ~none:`Insufficient_funds
```

Comment:
Not sure how much this matters but you could move this into `Ticket_map.update` with an exception to catch the `Insufficient_funds` case and turn into a result. Then you're only doing one lookup.

Reply:
If it were anywhere else in Deku I wouldn't even mention it, but you may care about that sort of thing for the VM, not sure.

Reply:
changed